solr/core/src/test/org/apache/solr/util/TestCircuitBreaker.java

```diff
@@ -157,6 +163,57 @@ public void testBuildingMemoryPressure() {
     }
   }
 
+  public void testFakeCPUCircuitBreaker() {
+    ExecutorService executor = ExecutorUtil.newMDCAwareCachedThreadPool(
+        new SolrNamedThreadFactory("TestCircuitBreaker"));
+    HashMap<String, String> args = new HashMap<String, String>();
+
+    args.put(QueryParsing.DEFTYPE, CircuitBreaker.NAME);
+    args.put(CommonParams.FL, "id");
+
+    AtomicInteger failureCount = new AtomicInteger();
+
+    try {
+      removeAllExistingCircuitBreakers();
+
+      CircuitBreaker circuitBreaker = new FakeCPUCircuitBreaker(h.getCore().getSolrConfig());
+
+      h.getCore().getCircuitBreakerManager().register(circuitBreaker);
+
+      for (int i = 0; i < 5; i++) {
+        executor.submit(() -> {
+          try {
+            h.query(req("name:\"john smith\""));
+          } catch (SolrException e) {
```

Comment:
Does circuit breaker throw a more specific exception? It's kind of cumbersome from a client API perspective that they need to inspect the string message for knowing that they should retry from circuit breakers.

Reply:
Shouldnt return code 500 help there?

Reply:
Currently the `CircuitBreakerManager.toErrorMessage` uses individual breaker's `getDebugInfo` to construct the error message. I think we should improve these individual messages to clearly say why a break was tripped. Currently it's not obvious - eg. when a mem breaker trips it just reports the current and total memory but not the thresholdPct.

Maybe we need a different method like `CircuitBreaker.getErrorMessage` ?

Reply:
return code 500 doesn't matter if the SolrJ client converts it into an exception every single time.

Reply:
https://issues.apache.org/jira/browse/SOLR-14755